qa/tasks/vstart_runner.py: remove the stale netnses and bridge

lxbsz opened the pull request:
If the previous test cases failed, the netnses maybe left and will
be stale for current test cases. Here will remove them when new
test begins.

Fixes: https://tracker.ceph.com/issues/45817
Signed-off-by: Xiubo Li <xiubli@redhat.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

mchangir commented on qa/tasks/vstart_runner.py:
```diff
@@ -1171,6 +1171,41 @@ def clear_old_log():
+    p = p.stdout.getvalue().strip()
+
+    # Get the netns name list
+    netns_list = re.findall(r'[^()\s][-.\w]+[^():\s]', p)
```
Do we need to worry about non-ceph netns ?
I know this script is strictly for ceph consumption, but this just popped up anyway.

lxbsz replied:
Hmm, I will fix and filter out the ceph related.
Thanks @mchangir

lxbsz commented:
> LGTM ... just a nit. Please check.

Fixed it. Thanks.

batrick commented on qa/tasks/cephfs/mount.py:
```diff
@@ -184,13 +184,19 @@ def _setup_netns(self):
-                if q is not None:
-                    found = True
-                    break
+                        "ip netns exec {0} ipd aaddr".format(ns_name)]
```
typo? `aaddr`??

lxbsz replied:
It seems the "Secure Shell App" editor's problem, sometime switching between chinese and english languages will add some extra chars I typed.

I will fix it.

batrick commented on qa/tasks/cephfs/mount.py:
```diff
@@ -184,13 +184,19 @@ def _setup_netns(self):
             for ns in netns_list:
                 ns_name = ns.split()[0]
                 args = ["sudo", "bash", "-c",
```
Is there a reason this is not:

`["sudo", "ip", "netns", ...]`

?

lxbsz replied:
No, just copied it from the mount.py code. I will fix it.

lxbsz replied:
For this, let keep it the same with others just like this ?   Or should we switch to "["sudo", "ip", "netns", ...]" ?

batrick replied:
Yes, please fix it.

lxbsz replied:
Okay, will fix them.

batrick commented on qa/tasks/vstart_runner.py:
```diff
@@ -1179,6 +1179,38 @@ def clear_old_log():
 
+# This will cleanup the stale netnses, which are from the
+# last failed test cases.
+def cleanup_stale_netnses_and_bridge():
```
Why not make this part of `qa/tasks/cephfs/mount.py` so it cleans up stale netns in teuthology too?

lxbsz replied:
IMO, it makes no sense to do this for the teuthology, because at beginning the test nodes will be rebooted, right ?  And all the stale netnses will be cleaned up after nodes rebooted. 

And also once current test case fail in any reason, no new test cases should be rerun and continued in teuthology.

batrick replied:
> And also once current test case fail in any reason, no new test cases should be rerun and continued in teuthology.

That may not always be the case.

In any case, I think we need to fix this from the other direction: `vstart_runner.py` should not be leaving these network bridges behind when it exits. I'm not sure why, but it does not appear to do any unmounts before exiting after completing the test. Can you try changing that instead?

lxbsz replied:
> > And also once current test case fail in any reason, no new test cases should be rerun and continued in teuthology.
> 
> That may not always be the case.
> 
> In any case, I think we need to fix this from the other direction: `vstart_runner.py` should not be leaving these network bridges behind when it exits. I'm not sure why, but it does not appear to do any unmounts before exiting after completing the test. Can you try changing that instead?

Normally if the test case finish without any assertion or errors, it will remove all the netnses and bridge, etc it created.  The problem is that if the test case is aborted with some errors, the vstart_runner.py will not have any chance to do the clean up.

lxbsz replied:
The stale netnses and bridge are all left from the previous failing test cases.

lxbsz replied:
The mount.py is not a good place to fix this for teuthology IMO, the mount.py just for each mountpoint, for example if there have 2 mountpoints, the second one may remove the first one's netns and the global bridge.

rishabh-d-dave commented on qa/tasks/cephfs/mount.py:
```diff
@@ -110,12 +132,13 @@ def _setup_brx_and_nat(self):
-            args = ["sudo", "bash", "-c", "ip link add name ceph-brx type bridge"]
+            args = ['sudo', 'ip', 'link', 'add', 'name', 'ceph-brx', 'type', 'bridge']
+            self.client_remote.run(args=args, timeout=(5*60))
+            args = ['sudo', 'ip', 'addr', 'flush', 'dev', 'ceph-brx']
             self.client_remote.run(args=args, timeout=(5*60))
```
Wouldn't it be bettter to do this in the teardown method?

lxbsz replied:
Do you mean the 'flush' here ?

rishabh-d-dave replied:
Yes.

lxbsz replied:
No, it won't work. 
This will trying to resolve the case:
When the test node was used to do some test cases and then after that the bridge will be removed, but the network config file will be left and if we reboot the node and run the test case again, when we setup the `ceph-brx` bridge again it will config it automatically by reading from the stale network config file.

Then if we config the bridge again it will fail.

lxbsz commented on qa/tasks/cephfs/mount.py:
```diff
@@ -16,6 +16,33 @@
+        args = ['sudo', 'ip', 'netns', 'delete', '{0}'.format(ns_name)]
+        try:
+            remote.run(args=args, timeout=(5*60))
+        except Exception:  
```
Will remove it.

lxbsz commented on qa/tasks/cephfs/mount.py:
```diff
@@ -16,6 +16,33 @@
+# This will cleanup the stale netnses, which are from the
+# last failed test cases.
+def cleanup_stale_netnses_and_bridge(remote):
+    p = remote.run(args=['ip', 'netns', 'list'],
+                   stderr=StringIO(), stdout=StringIO(),
+                   timeout=(5*60))
+    p = p.stdout.getvalue().strip()
+
+    # Get the netns name list
+    netns_list = re.findall(r'ceph-ns-[^()\s][-.\w]+[^():\s]', p)
```
Okay, will fix it.

rishabh-d-dave commented:
Tested this PR too, state netnses and bridges were deleted for me.

lxbsz commented:
Updated and fixed "TabError: inconsistent use of tabs and spaces in indentation
" in http://qa-proxy.ceph.com/teuthology/xiubli-2020-06-11_06:20:02-fs:basic_functional-wip-xiubli-fs-testing-2020-06-11-1206-distro-basic-smithi/5136215/teuthology.log

lxbsz commented:
Teuthology test passed:
http://pulpito.ceph.com/xiubli-2020-06-12_13:35:30-fs:basic_functional-wip-xiubli-fs-testing-2020-06-12-1613-distro-basic-smithi/

lxbsz commented:
The teuthology test passed:
http://pulpito.ceph.com/xiubli-2020-06-15_03:57:20-fs:basic_functional-wip-xiubli-fs-testing-2020-06-15-1001-distro-basic-smithi/

lxbsz commented:
Resolved the conflict with the omit_sudo changes.

lxbsz commented:
jenkins test make check

lxbsz commented:
jenkins retest this please

lxbsz commented:
jenkins test dashboard backend

rishabh-d-dave commented on qa/tasks/cephfs/mount.py:
```diff
@@ -14,6 +14,32 @@
+        ns_name = ns.split()[0]
+        args = ['sudo', 'ip', 'netns', 'delete', '{0}'.format(ns_name)]
+        try:
+            remote.run(args=args, timeout=(5*60))
```
Set `omit_sudo` to `True`.

rishabh-d-dave commented on qa/tasks/cephfs/mount.py:
```diff
@@ -14,6 +14,32 @@
+    # Remove the stale 'ceph-brx'
+    try:
+        args = ['sudo', 'ip', 'link', 'delete', 'ceph-brx']
+        remote.run(args=args, timeout=(5*60))
```
Set omit_sudo to True.

lxbsz replied:
You mean we should set it to False, right ?

We are fixing this in PR#35944 together with defer deleting the netns. In that PR we are using the run_shell_payload() and have fixed it :-)

Thanks @rishabh-d-dave

rishabh-d-dave replied:
Yes, I should've written "Set `omit_sudo` to `False`".

lxbsz commented:
Closing it since we are fixing this together with https://github.com/ceph/ceph/pull/35944.
